src/wrapper/strings/java_str.rs

```diff
@@ -27,11 +28,91 @@ impl<'a: 'b, 'b> JavaStr<'a, 'b> {
         Ok(java_str)
     }
 
-    /// Extract the raw C string pointer from the JavaStr. This will be
+    /// Get a pointer to the character array beneath a [JString]
+    /// The array contains Java's modified UTF-8: `CESU-8`
+    ///
+    /// # Warning
+    ///
+    /// The caller must release the array when they are done with it
+    // This function does not need to be unsafe,
+    // as a memory leak isn't considered unsafe nor is getting a raw pointer to something
+    fn get_string_utf_chars(obj: JString<'_>, env: &JNIEnv<'_>) -> Result<*const c_char> {
+        non_null!(obj, "get_string_utf_chars obj argument");
+        let ptr: *const c_char = jni_non_null_call!(
+            env.internal,
+            GetStringUTFChars,
+            obj.into_raw(),
+            ::std::ptr::null::<jboolean>() as *mut jboolean
+        );
+
+        Ok(ptr)
+    }
+
+    /// Get the raw C string pointer from the JavaStr. This will be
     /// encoded using the JVM internal `CESU-8`-style.
     pub fn get_raw(&self) -> *const c_char {
         self.internal
     }
+
+    /// Get the raw C string pointer from the [JavaStr]. This will be
+    /// encoded using the JVM internal `CESU-8`-style.
+    ///
+    /// # Warning
+    /// The programmer is responsible for making sure the backing string gets released
+    /// when they are done with it by resconstructing the [JavaStr] with [Self::from_raw],
+    /// which will release the backing string when it is dropped.
+    pub fn into_raw(self) -> *const c_char {
+        self.internal
+    }
+
+    /// Get a [JavaStr] from it's raw components
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that `ptr` is valid, and that `obj` is the same [JString] used when
+    /// creating the `ptr`
+    pub unsafe fn from_raw(ptr: *const c_char, obj: JString<'a>, env: &'b JNIEnv<'a>) -> Self {
+        Self {
+            internal: ptr,
+            obj,
+            env,
+        }
+    }
+
+    /// Release the backing string
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure [Self] was constructed from a valid pointer along with the same [JString]
+    /// as was originally used to get the pointer
+    ///
+    /// # Errors
+    ///
+    /// If the release call fails
+    pub unsafe fn release(mut self) -> Result<()> {
+        Self::release_string_utf_chars(&mut self)
+
+        // TODO do we even need this function? Or is it better to just use `Drop` for this?
+        // TODO do we need to use a ManuallyDrop here to avoid running the `Drop` code?
+    }
```

Comment:
Right, I don't think we should have this API. That would be a bit like if `Box` had a `free()` method that would free the backing memory which would leave it in an inconsistent state.

Reply:
Good point. I've removed it.